README
=============
Introduction:
=============

This distribution are designed to generate a freestanding, maintenance-free web
site for listing arXiv.org files (or other URLs), as seen at
http://astronomy.nmsu.edu/agso/coffee. 

A PHP-enabled web site lists submitted articles, has an input form for
submitting additional articles of interest, and upon submission the site
automatically updates the PHP code. Items can be subsequently deleted from the
list using a password-protected PHP text editor. 

Astroph.py was created at UC Los Angeles by Ian J. Crossfield and Nathaniel
Ross. 

It was then heavily edited by Ryan T. Hamilton at New Mexico State University,
and became this distributed version, available at
https://bitbucket.org/astrobokonon/astrocoffee/ 

Re-use or modification is allowed and encouraged, so long as proper
acknowledgement is made to the original authors and institutions.

Bug reports, suggestions, or just emailing to say that the code has been useful
are all greatly encouraged!

Ryan Hamilton
Astronomy Graduate Student
New Mexico State University
rthamilt@nmsu.edu
https://bitbucket.org/astrobokonon/astrocoffee/overview

=====================================
Installation of the astroph.py Suite:
=====================================

This will hopefully get you up and running with the code.  The instructions are
long.  Sorry.  Please follow along and you should be fine.  If all else fails,
you can email me!  It's likely that there are bugs, and I'll be happy to take a
look and see if the problem is on your end or in the code somewhere that needs
to be fixed.  

This will follow N steps; general requirements, changes to the .py files,
changes to the .php files, additional comments, and debugging hints.
 
=====================
General Requirements:
=====================

1) You need a web server with PHP and Python scripting enabled.  Plenty of web
resources exist to guide you through configuration of your particular web
server.  In addition, you need to have the 'dateutil' module available through
Python.  To check, start Python and type 

import datetime
from dateutil.relativedelta import relativedelta, MO, TU, WE, TH, FR, SA, SU

If either of these give an import error, contact your local system
administrator or check for a different version of Python to run.  It is not
uncommon to have many versions of Python living together on the same server for
compatibility with pyraf and the STSCI Python packages.  

This was built and tested using Python v2.5, v2.6, and v2.7.x; haven't tested
against Python 3.x at all.

2) Contrary to previous versions, BeautifulSoup 4.3.2 is now included.  It's
great.

=========================
Changes to the .py Files:
=========================

The following files should be edited by the user:

./Private/passwords.py
runcoffee.py

runcoffee.py
------------

You should first edit the runcoffee.py file to change the date and time of the
meeting, as well as who gets the error email when something goes wrong.  See
the 'day', 'hour', 'min', and 'responsibles' variables at the top of this file.

'day' is a two letter abbreviation of a weekday, where the abbreviations
(starting with Monday) are MO, TU, WE, TH, FR, SA, SU.  'hour' should be input
in 24h ('military time'), but are converted to the normal US format for display
purposes.

When errors are found, it will send an automated message to those listed in the
'responsibles' variable.  The contents of that message can be found in the file
error.mail.  If you'd like to change who the email appears to be coming from,
then change the 'whosend' variable to what you'd like.  This is the newest
section of the code and could need some adjustments, so bugs are likely.  It
might be necessary to change the 'server' variable to be whatever your mail
(SMTP) server should be.  See the send_error_mail.py file and the python smtp
documentation for more details/help.

If you'd like the ability to comment on each posting, then it is possible using
IntenseDebate.  Please follow these directions, which admitedly may have
changed since I first wrote this section of the code.

   --- Sign up at IntenseDebate.com
   --- Be sure to click the box to use IntenseDebate on my website/blog.  
   --- After verifying your email address, add your website address   
   --- Click on 'Generic Install'  
         All we need from the code is the value of 'var idcomments_acct=', 
   --- In runcoffee.py, change 'idcomments = False' to '= True'
   --- In runcoffee.py replace YOURHASHGOESHERE with that value and comment 
         out the line 'idcomments=False' line.  
         *** Be sure you have the single quote ticks (\') properly escaped ***
   --- At the top of the IntenseDebate page goto 'Sites', clicking on your 
         newly setup website, and then in the 'Settings->Misc' section 
   --- Set your comment location to 'Frontpage'.  

Then you can customize to your heart's content, including adding Facebook
Connect and logging in via OpenID and Twitter as well.

passwords.py
------------

One should change the username, password, and salt variables in the
passwords.py file in the ./Private directory, run it, and then copy and paste
the output into the listmanager.php file.  This insures a bit more security
than having the username and password in plaintext in the listmanager.php file,
though the salt is there for any enterprising young hacker.  

==========================
Changes to the .php Files:
==========================

The following files should be edited by the user:

index.php
listmanager.php

1) See the passwords.py section above and make the necessary changes in
listmanager.php

2) Determine the path to your python executable and set the $python variable in
both index.php and listmanager.php to point towards it.  This also allows you
to specify a particular version if multiple exist.  Typically the full path is
required, though not always.  If in doubt, ask your sysadmin.  Incorrect or
non-functional versions of this will cause the code to fail, silently.

3) Change your timezone in the index.php and listmanager.php files to match
your own.  The format for the timezone string can be found at the link in the
php file or be found via the googles.

4) You'll need to change the affiliation to your own institution/group in the
following files:

./index.php
./archive_top.php
./archive_bottom.php
./useful.links
./general.links

You can change the CSS to your own needs on your own time, if you're setting
this up you probably know more than enough to change the weather links and
insert a calendar or whatever else you need to do.  You can add in your own
customizations by changing the contents of the sidebars in the index.php file.
See <h2>Weather Checks</h2>, <h2>Useful Links</h2>, <h2>General Links</h2>, and
<h2>Calendar</h2> sections.  

Be careful to mind the <div> tags when adding/changing things since and
extra/missing ones will cause the page to look like crap and the CSS to go
nuts.

You probably can use a different CSS sheet, though I don't guarantee that it
will work very well without editing astroph.py to close the correct <div>'s.
Your Milage May Vary.

============
Final Steps:
============

1) Once you have everything the way you'd like it to be, execute the Perms.sh
file to make sure the permissions are set as they should.  It will attempt to
set the permissions necessary for you, but you should double check to see if
the choices make sense.  

The *.php files, papers, and dregs.log will all be continually updated, and so
need to have at least write access enabled.  You will want to update the *.txt
files to customize your page, but after that they only need read access
enabled, as do the *.py files.  ./Archive needs full read/write access as well.

Permission errors will cause the scripts to fail silently and not update
anything.

2) To control access and try to prevent spam, determine your CIDR address range
and change the htaccess file to match.  This will give a 403 Forbidden error to
any address not in the CIDR range, and is a quick and easy way to prevent spam
bots from polluting the page.  You can also add specific IPs to allow, though
this is a bit difficult unless you have known/static IP addresses.  

The htaccess file included in the tarball is an example of what I use.  IF YOU
CHOOSE TO USE THIS FEATURE, rename 'htaccess' to '.htaccess' after you
determine your CIDR range.

================
Debugging Hints:
================

At this point, everything should be set up.  Point your webbrowser towards
where you set things up, and try a valid submission.  Wait a (few) minutes and
refresh the page, and it all is well then you should be good to go!

If things have not updated, there are a few things to try:

   --- Check the directions again, making sure you have not missed any steps :)
   --- Check the file permissions.  Incorrect file permissions will cause the 
         code to fail silently.
   --- Check your python path/executable.  Incorrect values/versions will cause 
         the code to fail silently.
   --- Attempt to run the code as a (normal non-root) user.  In the directory, 
         'touch papers; python runcoffee.py' 
         and check for any exceptions/errors.
   --- Email me or ask your local computer or *nix guru for help

status.log and dregs.log can be very useful to see what's going on with the
code and should (hopefully) let you see how far things have progressed before
dying on you.

If all else fails, you can email me!  It's likely that there are bugs, and I'll
be happy to take a look and see if the problem is on your end or in the code
somewhere that needs to be fixed.  Bug reports, suggestions, or just emailing
to say that the code has been useful are all greatly encouraged!

Ryan Hamilton
Visiting PostDoctoral Scientist

HISTORY
-----------------------------
History:
-----------------------------

The following history is probably incomplete, since I stopped updating it when
I moved everything to the BitBucket Repository

2010-02-11 09:41 IJC: v0.4  -- First edition w/documentation.
2010-02-12 13:05 IJC: v0.41 -- Corrected "nexthursday" bug.
2010-04-04       RTH: v0.75 -- Added many features, including archiving and 
                                 ability to comment on papers via IntenseDebate
2010-04-09       RTH: v0.77 -- Fixed arXiv date scraping bug
2010-04-12       RTH: v0.85 -- Added ADS scraper
                                 Second edition w/documentation
2010-05-04       RTH: v0.90 -- Fixed Nature scraper due to website changes
                                 Improved exception handling to avoid making 
                                 the entire website useless when one preprint 
                                 fails for whatever reason
2010-05-09       RTH: v0.91 -- Added .htaccess file to restrict access and 
                                 hopefully cut out spam
2010-05-24       RTH: v0.92 -- Added xxx.lanl.gov server checking for arXiv
2010-07-19       RTH: v0.95 -- More exception handling to catch stray PDFs
                                also changed error reporting to log file.

------------------------------------
Comments removed from source code:
------------------------------------
General Comments

# 2010-02-05 15:24 IJC: v0.1: URLs are valid IDs.  "Next Thursday"
# automatically calculated.
# 2010-02-06 16:15 IJC: v0.2: Adding submission field.
# 2010-02-08 18:04 IJC: v0.21: Added logfile
# 2010-02-09 11:08 IJC: v0.3: Using text files for header/footer
# 2010-02-11 09:30 IJC: v0.4: Using PHP article list manager.
# 2010-02-12 11:45 IJC: v0.41: Using multiple PHP headers (for day-of-week)
# 2010-02-21 17:00 RTH: v0.42: Started NMSU customizations 
# 2010-02-22 12:50 RTH: v0.43: Works, shortened time, added refresh when 
#                               changing submission file
# 2010-03-01 01:00 RTH: v0.50: Major changes & style updates
# 2010-03-06 13:00 RTH: v0.55: Major changes to calculation of meeting dates
#                               added auto-archive functionality
# 2010-03-16 19:00 RTH: v0.56: Fixed bug in next meeting calculation, added
#                               passing of meeting date info for simplicity
#                               as well as direct links for arXiv submissions
# 2010-03-20 21:00 RTH: v0.60: Implimented preprint class type constructor
# 2010-03-22 16:00 RTH: v0.65: Full archive capabilities, code cleanup
# 2010-03-23 19:00 RTH: v0.67: Got IntenseDebate comments working
# 2010-04-02       RTH: v0.75: Major code organization/cleanup, including using
#                               BeautifulSoup to parse pages instead of reg exps
#                               Also added IntenseDebate id input for easy setup
# 2010-04-06       RTH: v0.77: Bugfix, changed handling of 'www' links and 
#                               arxiv date scraping
# 2010-04-12       RTH: v0.85: Added ADS scraper, works same as the others :)
# 2010-04-29       RTH: v0.88: Fixed Nature scraper
# 2010-05-04       RTH: v0.90: Much better exception handling; one failure to 
#                               find a particular element of the preprint object
#                               shouldn't bring down the whole page (I hope)
# 2010-05-24       RTH: v0.92: Added xxx.lanl.gov server checking for arXiv
# 2010-07-19       RTH: v0.95: More exception handling to catch stray PDFs
#                               also changed error reporting to log file.

getunique
# 2010-04-06 RTH: Created, though just copied and pasted from
#                   http://www.peterbe.com/plog/uniqifiers-benchmark

getinfo
# 2010-02-03 09:38 IJC: Created
# 2010-02-05 15:08 IJC: Now gets title of HTML pages, too!  And
# corrects "author" links.
# 2010-02-09 11:05 IJC: Now do a better check for arxiv-IDs and valid URLs
# 2010-04-02       RTH: Change to better detect astro-ph types
# 2010-04-06       RTH: Kludge to pass error info about each submission
# 2010-04-12       RTH: Added ADS scraper

readlist
# 2010-02-03 10:06 IJC: Created
# 2010-02-05 11:24 IJC: Added 'sleep' command

getnextduedate
# 2010-03-05 RTH: Created
# 2010-03-16 RTH: Fixed bug where next meeting would be 2 weeks away, added
#                  passing of meeting info for easier use
# 2010-04-02 RTH: Changed status message to be passed back to main function

getprevduedate
# 2010-03-05 RTH: Created
# 2010-04-02 RTH: Changed status message to be passed back to main function

makeheader
# 2010-02-12 11:45 IJC: Now get correct Thursday
# 2010-03-05 RTH: Get next date in a much easier way

makefooter
# 2010-02-09 10:29 IJC: Updated to use coffee_footer_php.txt
# 2010-04-09       RTH: Removed that dependence since I don't use it

makehtml
# 2010-02-03 10:13 IJC: Created

astroph.py
#!/usr/bin/env python
"""
A module to generate HTML code using specified ArXiV file IDs.

The main function is:
   docoffee -- read in a file of arxiv IDs or interesting URLs, and
               write HTML code to specified file.

There should be no need to edit this file or any of the asssociated parsing
functions (get*.py). There is a wrapper script, runcoffee.py, that does most
of the front-end organization and contains the user changable variables.

Initially created on 2010-02-05 at UCLA by Ian J. Crossfield and Nathaniel
Ross. Modified by Ryan T. Hamilton at New Mexico State University.

This version under development by Abhimat K. Gautam at UCLA.

Contribute or fork on GitHub at https://github.com/abhimat/AstroCoffee
"""
# 2010-10-20       RTH: v0.98: Spun off parsers into own files.  Removed all
#                                changes to the CHANGELOG/online repository

__version__ = '0.98.9'


class preprint(object):
    """Empty object container for preprints"""
    def __init__(self):
        self.url = ''
        self.author = ''
        self.numauth = ''
        self.title = ''
        self.date = ''
        self.abstract = ''
        self.subject = ''
        self.comments = ''
        self.sources = ''
        self.commentid = ''
        self.errors = '0'


def getunique(seq):
    """Take an input list and return list of unique elements, order preserved
    """
    checked = []
    for e in seq:
        if e not in checked:
            checked.append(e)
    return checked


def getinfo(id, server='http://arxiv.org/abs/'):
    """Take an ArXiV ID and return the title, authors, abstract, sub. date

    INPUT:
       id -- (str): ArXiV ID (e.g. '1002.0504v1' or 1002.0504)
              _OR_
             (str) URL, from which TITLE tag will be extracted.

    OUTPUT:
       a preprint-class object.
    """
    from getadsinfo import getadsinfo
    from getarxivinfo import getarxivinfo
    from getvixrainfo import getvixrainfo
    from getnatureinfo import getnatureinfo
    from getscienceinfo import getscienceinfo
    from getvoxchartainfo import getvoxchartainfo
    from getwebinfo import getwebinfo
    import urllib2

    servererr = False
    # Set up the ID
    id = str(id).strip().lower()

    # Check for the various types of arXiv identifiers
    try:
        # Check for plain numbers with no lead/trail and a period in it
        if (id.find('.') > -1):
            idnum = float(id[0:9])
            isValidArxiv = len(id) >= 9
            # Ok, it's just a number so put the proper url stuff in front of it
            id = server + id
            servererr = False
    except:
        # Ok, wasn't just numbers.  Look for arxiv:###... style, but not
        #   valid arxiv.org addresses that were submitted
        if (id.find('arxiv')>-1 or id.find('arXiv')>-1) and \
           (id.find('.org') == -1) and \
           (id.find('.gov') == -1):
            isValidArxiv = True
            id = server + id
        elif (id.find('astro-ph') > -1):
            isValidArxiv = True
            id = 'http://arxiv.org/abs/' + id
        else:
            isValidArxiv = False
    
    # HTTP request for a webpage URL
    try:
        ## Add headers to HTTP request so don't get 403 error
        hdr = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11',
               'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
               'Accept-Charset': 'ISO-8859-1,utf-8;q=0.7,*;q=0.3',
               'Accept-Encoding': 'none',
               'Accept-Language': 'en-US,en;q=0.8',
               'Connection': 'keep-alive'}
        request = urllib2.Request(id, headers=hdr)
        
        html = urllib2.urlopen(request).read()
        urlpage = id  # For compatibility down lower in the code
    except urllib2.HTTPError, e:
        print e.code
        html = e.read()
        urlpage = id
        servererr = True
    except:
        # Hm...didn't open, and not a 404 or something; try adding http://
        if (id.startswith('http://') is False):
            urlpage = 'http://' + id
            try:
                html = urllib2.urlopen(urlpage).read()
            except:
                # Ok...try adding www. if it's not there
                if (urlpage.startswith('http://www.') is False):
                    urlpage = 'http://www.' + id
                try:
                    html = urllib2.urlopen(urlpage).read()
                except:
                    html = '<html><head>' + \
                           '<title>BAD LINK: ' + id + \
                           '</title>' + \
                           '</head><body></body></html>'
                    servererr = True
        else:
            html = '<html><head>' + \
                   '<title>BAD LINK: ' + id + \
                   '</title>' + \
                   '</head><body></body></html>'
    
    if servererr == False:
        # Nature Article
        if urlpage.find('nature.com')>-1:
            if urlpage.find('nature.com/news/')>-1:
                thispaper = getwebinfo(urlpage, html)
                if thispaper.errors == '0':
                    thispaper.errors = 'Success reading ' + urlpage
                else:
                    thispaper.errors = 'Some ERRORS reading ' + urlpage
                thispaper.id = ''
            else:
                thispaper = getnatureinfo(urlpage, html)
                if thispaper.errors == '0':
                    thispaper.errors = 'Success reading ' + urlpage
                else:
                    thispaper.errors = 'Some ERRORS reading ' + urlpage
                thispaper.id = ''
        # ADS Result
        elif urlpage.find('adsabs.harvard.edu')>-1:
            thispaper = getadsinfo(urlpage, html)
            if thispaper.errors == '0':
                thispaper.errors = 'Success reading ' + urlpage
            else:
                thispaper.errors = 'Some ERRORS reading ' + urlpage
            thispaper.id = ''
        # arXiv Article
        elif urlpage.find('arxiv.org')>-1    or \
             urlpage.find('xxx.lanl.gov')>-1 or \
             (isValidArxiv): 
            thispaper = getarxivinfo(urlpage, html)
            if thispaper.errors == '0':
                thispaper.errors = 'Success reading ' + urlpage
            else:
                thispaper.errors = 'Some ERRORS reading ' + urlpage
            thispaper.id = ''
        # Science Article
        elif urlpage.find('sciencemag.org')>-1:
            thispaper = getscienceinfo(urlpage, html)
            if thispaper.errors == '0':
                thispaper.errors = 'Success reading ' + urlpage
            else:
                thispaper.errors = 'Some ERRORS reading ' + urlpage
            thispaper.id = ''
        # Vox Charta Article
        elif urlpage.find('voxcharta.org')>-1: 
            thispaper = getvoxchartainfo(urlpage, html)
            if thispaper.errors == '0':
                thispaper.errors = 'Success reading ' + urlpage
            else:
                thispaper.errors = 'Some ERRORS reading ' + urlpage
            thispaper.id = ''
        # viXra Article
        elif urlpage.find('vixra.org')>-1: 
            thispaper = getvixrainfo(urlpage, html)
            if thispaper.errors == '0':
                thispaper.errors = 'Success reading ' + urlpage
            else:
                thispaper.errors = 'Some ERRORS reading ' + urlpage
            thispaper.id = ''
        # Anything Else
        else: 
            thispaper = getwebinfo(urlpage, html)
            if thispaper.errors == '0':
                thispaper.errors = 'Success reading ' + urlpage
            else:
                thispaper.errors = 'Some ERRORS reading ' + urlpage
            thispaper.id = ''
    # Error Handler
    else:
        thispaper = getwebinfo(urlpage, html)
        thispaper.errors = 'ERRORS reading ' + id
        thispaper.id = ''
        thispaper.url = id
    
    return thispaper


def readlist(filelist, sleep=60):
    """Get data for all papers with IDs or URLs specified.

    INPUT:
        filelist:  one of the following:
           (list) python list of IDs for astroph.getinfo
           (str) filename of an ascii file containing IDs

    OPTIONAL INPUT:
        sleep -- (float) number of seconds to wait between URL
                 requests, so aRxIv doesn't think you're a robot and
                 ban your connection.

    OUTPUT:
        a list of preprint-class objects
    """
    import time
    papers = []
    errors = ''

    if isinstance(filelist, list):
        ids = filelist
    else:  # Must be the name of a file to read:
        try:
            f = open(filelist, 'r')
            ids = f.readlines()
            f.close()
        except:
            errors = "Could not open file: %s" % filelist
#            print "Could not open file: %s" % filelist
            ids = ['']

    for id in ids:
        if len(id) > 5:
            thispaper = getinfo(id)
            errors = errors + thispaper.errors + "\n"
            if isinstance(thispaper, preprint):
                papers.append(thispaper)
                time.sleep(sleep)

    return errors, papers


def getnextduedate(d, h, m):
    """Get the next due date for meeting.

    INPUT:
        d:  datetime object day of the week to meet
        h:  datetime object hour of day to meet (24h)
        m:  datetime object minute of hour to meet

    OUTPUT:
        datetime of next meeting
    """

    import datetime
    from dateutil.relativedelta import relativedelta, \
        MO, TU, WE, TH, FR, SA, SU

    today = datetime.datetime.now()
    nextmeet1 = today+relativedelta(weeks=+1, weekday=d, hour=h,
                                    minute=m+35, second=0, microsecond=0)
    nextmeet2 = today+relativedelta(weekday=d, hour=h, minute=m+35, second=0,
                                    microsecond=0)
    msg = ''

    if abs(nextmeet2.day-today.day) < 1 and today < nextmeet2:
        nextmeet = nextmeet2
        msg = "Less than one day difference, so the next meeting is today"
    elif today < nextmeet2:
        nextmeet = nextmeet2
    else:
        nextmeet = nextmeet1

    return msg, nextmeet


def getprevduedate(d, h, m):
    """Get the previous date for meeting.

    INPUT:
        d:  datetime object day of the week to meet
        h:  datetime object hour of day to meet (24h)
        m:  datetime object minute of hour to meet

    OUTPUT:
        datetime of the previous meeting
    """

    import datetime
    from dateutil.relativedelta import relativedelta, \
        MO, TU, WE, TH, FR, SA, SU

    # Get current date
    today = datetime.datetime.now()
    prevmeet1 = today+relativedelta(weeks=-1, weekday=d, hour=h, minute=m+35,
                                    second=0, microsecond=0)
    prevmeet2 = today+relativedelta(weekday=d, hour=h, minute=m+35, second=0,
                                    microsecond=0)
    msg = ''

#    print prevmeet1, prevmeet2

    if abs(prevmeet2.day-today.day) < 1 and today > prevmeet2:
        prevmeet = prevmeet2
        msg = "Less than one day difference, so the last meeting was today"
    else:
        prevmeet = prevmeet1

    return msg, prevmeet


def makeheader(day, hour, min, php=False):
    """Return the HTML header info.  If specified, use PHP and submission form.
    """

    import datetime
    from dateutil.relativedelta import relativedelta, \
        MO, TU, WE, TH, FR, SA, SU

    # Need this dummy return value because of the way I write the status file
    mgs = ''
    msg, nextdate = getnextduedate(day, hour, min)#+relativedelta(minutes=-35)
    nextdate = nextdate+relativedelta(minutes=-35)
    datestr = nextdate.strftime('%a, %b %d, %Y at %I:%M %p')
    
    
    # If there are two meetings, use following
    ## Get previous date astro-ph coffee is held in week
    nextdate_early = nextdate+relativedelta(days=-2, minutes=+60)
    datestr_early = nextdate_early.strftime('%a, %b %d, %Y at %I:%M %p')

    head = []

    titleString1 = '<article class="block">'
    
    titleString2 = '<center><p>Suggested papers for<br><strong>{0}</strong></p></center>\n'.format(datestr)
    
    ## For two meetings use following instead
    titleString2 = '<center><p>Suggested papers for<br><strong>{0}</strong> and <strong>{1}</strong></p></center>\n'.format(datestr_early, datestr)
    head.append(titleString1)
    head.append(titleString2)

    # head.append('<p>astro-ph coffee is held Mondays at 1:30 PM and Wednesdays at 2:00 PM.')
    # head.append('Some suggested papers of interest are listed below.')
    # head.append('</p></div>')
    head.append('</article>\n')

    return head


def makefooter(php=False):
    """
    Return the HTML footer (list of str).  If called, append PHP scripting.
    """

    from time import localtime
    f = []
    yr, mo, day, hr, min, sec, wd, yd, ii = localtime()
    f.append("<article class=\"block\">")
    f.append('<hr><center><p class="small">')
    # f.append("<a href='http://astronomy.nmsu.edu/rthamilt/astrocoffee/'>")
    # f.append("astroph.py v%s</a> based on the " % __version__)
    # f.append("<a href='http://www.astro.ucla.edu/~ianc/astroph.shtml'>")
    # f.append("original astroph.py</a>.  ")
    f.append("Updated %i/%02i/%02i %02i:%02i:%02i</p></center>" %
            (yr, mo, day, hr, min, sec))
    f.append("</article>")
    
    foot = [line+'\n' for line in f]

    return foot


def makehtml(papers, day, hour, min, idcomments=False, php=False):
    """Return HTML code for a public ArXiV web page from paper list.

    INPUT:
       papers -- (list) python list of preprint-class objects.

    OPTIONAL INPUT:
       php -- (bool) whether to generate HTML containing PHP scripting
                for paper submissions

    OUTPUT:
       html -- HTML code for generating a web page
    """

    import time
    import re

    # Generate header
    h = makeheader(day, hour, min, php=php)
    # h = ['<p>Test!</p>\n']

    # generate footer
    f = makefooter(php=php)

    # Generate text from preprints
    body = []
    date = ''
    for paper in papers[::-1]:
        date = paper.date
        if isinstance(paper, preprint):
            if paper.errors.startswith("Success"):
                if paper.sources != '':
                    if date != "":
                        body.append('<article class="block">')
                        body.append('<div class="date small">%s</div>' % paper.date)
                    # Remove any stray/extra whitespaces
                    paper.sources = paper.sources.lstrip()
                    paper.sources = paper.sources.rstrip()
                    body.append('<div class="links small">[ %s ]</div>' %
                                paper.sources)
                    title = '<h3><a href="%s">%s</a></h3>' \
                            % (paper.url, paper.title)
                else:
                    if date != "":
                        body.append('<article class="block">')
                        body.append('<div class="date small">%s</div>' % paper.date)
                    title = '<h3><a href="%s">%s</a></h3>' % (paper.url, paper.title)
                body.append('%s' % title)
                if paper.numauth > 5 and \
                   paper.author != "Error Grabbing Authors":
                    authremain = paper.numauth-5
                    aexstring = ", + " + str(authremain) + " more"
                    paper.author = paper.author + aexstring
                body.append('<div class="authors small">%s</div>' % paper.author)
                # Limit the length of the abstract displayed, and if its
                #   shorter then just display the whole thing
                abslength = 500
                # Hopefully remove links and other HTML things in there
                if len(paper.abstract) > abslength:
                    paper.shortabs = paper.abstract[0:abslength] + '...'
                    paper.shortabs = paper.shortabs.rstrip()
                else:
                    paper.shortabs = paper.abstract
                paper.shortabs = paper.shortabs.lstrip()
                paper.shortabs = paper.shortabs.rstrip()
                body.append('<div id="abstract"><p>%s</p></div>' % paper.shortabs)
                body.append('</article>')
            else:
                body.append('<article class="block">')
                if (paper.url.find('.pdf') > -1):
                    paper.date = "Please Do NOT Submit Direct PDF Links:"
                else:
                    paper.date = "Unknown Submission/Link:"
                body.append('<div class="date small">%s</div>' % paper.date)
                title = '<h3><a href="%s">%s</a></h3>' % (paper.url, paper.url)
                body.append('%s' % title)
                body.append('</article>')

    body = [line+'\n' for line in body]

    # Concatenate everything together
    html = h + body + f

    return html


def doarchivepage(file, outfile):
    """Create the archive file, clear the papers file

    INPUTS:
       file: (str) path to the papers file
       outfile: (str) path to archive file to create
    """

    import glob
    import shutil
    import re

    arcstat = ''

    try:
        f = open(outfile)
        arcstat = "Old papers already archived\n\n"
        f.close()
    except:
        arcstat = "Past previous deadline, so archiving old papers\n\n"

        # Easy file concatenation:
        f = open(outfile, 'wb')
        shutil.copyfileobj(open('./archive_top.php', 'rb'), f)
        shutil.copyfileobj(open('./astro_coffee.php', 'rb'), f)
        shutil.copyfileobj(open('./archive_bottom.php', 'rb'), f)
        f.close()

        # Get a list of all the current archived papers
        archived = glob.glob('./archive/*papers.php')
        # Now make the index page for archive, the long way since we add html
        page = open('./archive_top.php', 'r').read()
        page = page + '                <ul style="columns: 2; list-style-type: none;">\n'
        arclinks = ''
        for files in sorted(archived):
            files = re.sub(r'\./archive\/', '', files)
            linkname = re.sub(r'[\.\-]papers\.php', '', files)
            arclinks = arclinks + '                    <li><a href="' + files + '">' + \
                       linkname + '</a></li>\n'
        page = page + arclinks
        page = page + '</ul>\n'
        page = page + open('./archive_bottom.php', 'r').read()
        # Append it to our output file
        outfile = './archive/index.php'
        f = open(outfile, 'w')
        f.write(page)
        f.close()

        try:
            # Clear the papers file for the new week, but save the last
            # entry because it was what triggered the update/archive
            f = open(file, 'r')
            ids = f.readlines()
            f.close()

            f = open(file, 'w')
            f.write(ids[len(ids)-1])
            f.close()
        except:
            html = ['Could not get the papers info in the archive loop']
            papers = []
            arcstat = arcstat + "Could not get paper info in archive loop\n\n"
    return arcstat


def docoffeepage(file, url, day, hour, min, sleep=60, idid=False, php=False):
    """Read in arxiv IDs from a specified ASCII file, and write HTML.

    INPUTS:
       file: (str) path of ASCII file with arxiv IDs
       url:  (str) path of HTML file to write
       day:  (dateutil) day of meeting time
       hour: (flt) hour of meeting time
       min:  (flt) minute of meeting time

    OPTIONAL INPUT:
        sleep -- (float) number of seconds to wait between URL
                 requests, so aRxIv doesn't think you're a robot and
                 ban your connection.
    """

    import datetime

    today = datetime.datetime.now()
    msg1, prevdate = getprevduedate(day, hour, min)
    msg2, nextdate = getnextduedate(day, hour, min)

    outstat = msg1 + "\n" + msg2
    outstat = outstat + "\nPrevious: " + prevdate.strftime('%Y-%m-%d')
    outstat = outstat + "\nNow:      " + today.strftime('%Y-%m-%d %H:%M:%S')
    outstat = outstat + "\nNext:     " + nextdate.strftime('%Y-%m-%d') + "\n\n"

    outfile='./archive/' + prevdate.strftime('%Y-%m-%d') + '-papers.php'
    arcstat = doarchivepage(file, outfile)
    paperrs = ''

    # Ok, now read the papers for real
    try:  # Read in the papers
        paperrs, papers = readlist(file, sleep=sleep)
        html = []
    except Exception, why:
        html = ['Could not get the papers info']
        papers = []
        outstat = outstat + 'Could not get the paper info\n\n'
        outstat += str(why) + "\n\n"
    
    try:
        html = makehtml(papers, day, hour, min, idcomments=idid, php=php)
    except:
        html.append('Could not generate HTML')
        outstat = outstat + 'Could not generate HTML code\n\n'

    # Write the file
    try:
        # with open(url, 'w') as f:
        #           f.writelines(html)
        f = open(url, 'w')
        f.writelines(html)
        f.close()
    except:
        outstat = outstat + 'Could not write HTML code to file ' + url + "\n\n"

    outstat = str(outstat) + str(arcstat) + str(paperrs)
    return (html, outstat)
